RealmSwift/Handover.swift

```diff
@@ -0,0 +1,181 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Realm
+
+#if swift(>=3.0)
+    
+/// A Realm-bound object that can only be passed between threads by exporting for handover
+@objc public protocol ThreadConfined {
```

Comment:
From a Swift Evolution [thread](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20160815/026366.html),

> The implementation for self-conforming opaque protocols is not difficult, because the value itself can already be of any size, so it’s really not a problem to have an existential in there. In theory, someone could cook it up in a week or so.
> 
> For class protocols, I don’t know how to do it without an efficiency hit unfortunately.
>
> Consider these two functions, taking a homogeneous and heterogeneous array of a class-bound protocol type:
> 
> protocol P : class {}
> 
> func f<T : P>(array: [T]) {} // this takes an array of pointers to T, because there’s only one witness table for all of them
> func ff(array: [P]) {} // this takes an array of <T, witness table> pairs, two pointers each, because each element can be a different concrete type
> 
> What you’re saying is that f() should in fact allow both representations, because you’ll be able to call f() with a value of type [P]. Right now, if we know a generic parameter is class-constrained, we use a much more efficient representation for values of that type, that is known to be fixed size in the LLVM IR. We would have to give that up to allow class-constrained existentials to self-conform, since now a class-constrained parameter can be an existential with any number of witness tables.
> 
> There might be some trick for doing this efficiently, but I don’t know of one yet.
> 
> Of course, we can just say that class-constrained protocols never self-conform, unless they’re @objc. That seems like a hell of an esoteric restriction though (can you imagine trying to come up with a clear phrasing for *that* diagnostic?)
> 
> And if you’re wondering, the reason that @objc protocols self-conform in Swift today, is because they their existentials don’t have *any* witness tables — @objc protocol method bodies are found by looking inside the instance itself.
> 
> AnyObject is the other kind of protocol that self-conforms — you can use it both as a generic constraint, and as a concrete type bound to a generic parameter, and it ‘just works’, because again it doesn’t have a witness table.

-Slava Pestov